File: backend/src/application/services/webhook.service.ts
import { Injectable } from '@nestjs/common';
import { SaleRepository } from '../../infrastructure/database/repositories/sale.repository';
import { ProductRepository } from '../../infrastructure/database/repositories/product.repository';
import { UserRepository } from '../../infrastructure/database/repositories/user.repository';
import { SettingsRepository } from '../../infrastructure/database/repositories/settings.repository';
import { CouponRepository } from '../../infrastructure/database/repositories/coupon.repository';
import { MercadoPagoService } from './mercadopago.service';
import { EfiService } from './efi.service';
import { DeliveryService } from './delivery.service';
import { createHmac } from 'crypto';

@Injectable()
export class WebhookService {
  constructor(
    private readonly saleRepo: SaleRepository,
    private readonly productRepo: ProductRepository,
    private readonly mpService: MercadoPagoService,
    private readonly efiService: EfiService,
    private readonly deliveryService: DeliveryService,
    private readonly settingsRepo: SettingsRepository,
    private readonly couponRepo: CouponRepository,
  ) {}

  async processMercadoPago(body: any, headers?: any, query?: any): Promise<void> {
    console.log('[Webhook MP] Recebido - body:', JSON.stringify(body));
    console.log('[Webhook MP] Query:', JSON.stringify(query));

    const xSignature = headers?.['x-signature'];
    const xRequestId = headers?.['x-request-id'];
    const settings = await this.settingsRepo.get();
    const secret = settings?.mercadopagoConfig?.webhook_secret;

    if (secret && xSignature && xRequestId) {
      const parts = xSignature.split(',');
      let ts = '';
      let v1 = '';
      for (const part of parts) {
        const [key, val] = part.split('=');
        if (key === 'ts') ts = val;
        if (key === 'v1') v1 = val;
      }
      const dataId = query?.['data.id'] || body?.data?.id || body?.id;

      if (ts && v1 && dataId) {
        const manifest = `id:${dataId};request-id:${xRequestId};ts:${ts};`;
        const hash = createHmac('sha256', secret).update(manifest).digest('hex');
        if (hash !== v1) {
          console.error('[Webhook MP] Assinatura inválida');
          throw new Error('Assinatura do webhook inválida');
        }
        console.log('[Webhook MP] Assinatura válida');
      }
    }

    const paymentId =
      query?.['data.id'] ||
      body?.data?.id ||
      body?.id ||
      body?.resource?.split('/').pop();

    if (!paymentId) {
      console.log('[Webhook MP] Nenhum paymentId identificado, ignorando');
      return;
    }

    console.log(`[Webhook MP] Processando paymentId: ${paymentId}`);

    let sale = await this.saleRepo.findByPaymentId(paymentId.toString());

    if (!sale) {
      try {
        const mpPayment = await this.mpService.getPaymentStatus(paymentId.toString());
        if (mpPayment?.external_reference) {
          sale = await this.saleRepo.findById(mpPayment.external_reference);
          console.log(`[Webhook MP] Sale encontrada via external_reference: ${mpPayment.external_reference}`);
        }
      } catch (err: any) {
        console.error(`[Webhook MP] Erro ao buscar pagamento no MP:`, err?.message);
      }
    }

    if (!sale) {
      console.error(`[Webhook MP] Sale não encontrada para payment ${paymentId}`);
      return;
    }

    if (sale.status === 'approved') {
      console.log(`[Webhook MP] Sale ${sale.id} já aprovada, ignorando`);
      return;
    }

    const mpPayment = await this.mpService.getPaymentStatus(paymentId.toString());
    if (!mpPayment) {
      console.error(`[Webhook MP] Não foi possível obter status do payment ${paymentId}`);
      return;
    }

    console.log(`[Webhook MP] Status no MP: ${mpPayment.status} | Sale atual: ${sale.status}`);

    const newStatus =
      mpPayment.status === 'approved'
        ? 'approved'
        : mpPayment.status === 'rejected' || mpPayment.status === 'cancelled'
          ? 'failed'
          : 'pending';

    if (sale.status !== newStatus) {
      await this.saleRepo.update(sale.id, {
        status: newStatus,
        webhookVerified: true,
        paymentDetails: {
          ...sale.paymentDetails,
          paymentId: paymentId.toString(),
          status: mpPayment.status,
        },
      });

      console.log(`[Webhook MP] Sale ${sale.id} atualizada para ${newStatus}`);

      if (newStatus === 'approved') {
        if (sale.couponCode) {
          const coupon = await this.couponRepo.findByCode(sale.couponCode);
          if (coupon) {
            await this.couponRepo.incrementUses(coupon.id);
            console.log(`[Webhook MP] Cupom ${sale.couponCode} uso incrementado`);
          }
        }
        await this.deliveryService.deliver(sale.id);
      } else if (newStatus === 'failed') {
        await this.releaseStockIfNeeded(sale.productId);
      }
    }
  }


  async processEfi(body: any, headers?: any): Promise<void> {
    if (headers) {
      const clientVerify = headers['x-ssl-client-verify'] || headers['ssl-client-verify'];
      if (clientVerify && clientVerify !== 'SUCCESS') {
        throw new Error('Falha na validação mTLS da Efí');
      }
    }

    const pix = body.pix;
    if (!pix || !Array.isArray(pix) || pix.length === 0) return;

    for (const pixItem of pix) {
      const txid = pixItem.txid;
      if (!txid) continue;

      const sale = await this.saleRepo.findByTxid(txid);
      if (!sale) {
        console.error(`[Webhook Efí] Sale não encontrada para txid ${txid}`);
        continue;
      }

      if (sale.status === 'approved') continue;

      try {
        const charge = await this.efiService.getPixCharge(txid);
        const isPaid = charge.status === 'CONCLUIDA';

        if (isPaid && sale.status !== 'approved') {
          await this.saleRepo.update(sale.id, {
            status: 'approved',
            webhookVerified: true,
            paymentDetails: {
              ...sale.paymentDetails,
              status: 'CONCLUIDA',
            },
          });

          if (sale.couponCode) {
            const coupon = await this.couponRepo.findByCode(sale.couponCode);
            if (coupon) {
              await this.couponRepo.incrementUses(coupon.id);
              console.log(`[Webhook Efí] Cupom ${sale.couponCode} uso incrementado`);
            }
          }

          await this.deliveryService.deliver(sale.id);
        }
      } catch (error: any) {
        console.error(`[Webhook Efí] Erro ao verificar cobrança ${txid}:`, error?.message);
      }
    }
  }

  private async releaseStockIfNeeded(productId: string) {
    const product = await this.productRepo.findById(productId);
    if (product && product.stock !== null && product.stock !== undefined) {
      await this.productRepo.releaseStock(productId, 1);
    }
  }
}

File: backend-go/services/webhook_service.go
package services

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"strings"

	"github.com/jackc/pgx/v5/pgxpool"
	"vematize-backend-go/db"
)

type MPWebhookBody struct {
	ID    interface{}        `json:"id"`
	Type  string             `json:"type"`
	Topic string             `json:"topic"`
	Data  *MPWebhookBodyData `json:"data"`
}

type MPWebhookBodyData struct {
	ID string `json:"id"`
}

type EfiWebhookBody struct {
	Pix []EfiWebhookPixItem `json:"pix"`
}

type EfiWebhookPixItem struct {
	Txid string `json:"txid"`
}

func ValidateMPSignature(secret string, xSignature, xRequestId, dataID string) bool {
	log.Printf("[Debug] ValidateMPSignature acionado. xRequestId: %s, dataID: %s", xRequestId, dataID)

	if secret == "" || xSignature == "" || xRequestId == "" || dataID == "" {
		log.Println("[Debug] Elementos vazios para validacao de assinatura do MercadoPago")
		return false
	}

	parts := strings.Split(xSignature, ",")
	var ts string
	var v1 string
	for _, part := range parts {
		subParts := strings.Split(part, "=")
		if len(subParts) != 2 {
			continue
		}
		k := strings.TrimSpace(subParts[0])
		v := strings.TrimSpace(subParts[1])
		if k == "ts" {
			ts = v
		} else if k == "v1" {
			v1 = v
		}
	}

	if ts == "" || v1 == "" {
		log.Println("[Debug] Assinatura sem ts ou v1")
		return false
	}

	manifest := fmt.Sprintf("id:%s;request-id:%s;ts:%s;", dataID, xRequestId, ts)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(manifest))
	expectedHash := hex.EncodeToString(mac.Sum(nil))

	valid := hmac.Equal([]byte(expectedHash), []byte(v1))
	log.Printf("[Debug] Validacao de assinatura concluida. Resultado: %t", valid)
	return valid
}

func ProcessMercadoPago(pool *pgxpool.Pool, bodyBytes []byte, headers map[string]string, queryParams map[string]string) error {
	log.Println("[Debug] ProcessMercadoPago acionado no Go")

	if pool == nil {
		log.Println("[Debug] Erro: Pool de banco de dados nulo no ProcessMercadoPago")
		return errors.New("pool de banco de dados nulo")
	}

	var body MPWebhookBody
	if err := json.Unmarshal(bodyBytes, &body); err != nil {
		log.Printf("[Debug] Erro ao decodificar payload do webhook MP: %v", err)
		return err
	}

	xSignature := headers["x-signature"]
	xRequestId := headers["x-request-id"]

	var dataID string
	if body.Data != nil && body.Data.ID != "" {
		dataID = body.Data.ID
	} else if body.ID != nil {
		dataID = fmt.Sprintf("%v", body.ID)
	} else if queryID, exists := queryParams["data.id"]; exists {
		dataID = queryID
	} else if queryID, exists := queryParams["id"]; exists {
		dataID = queryID
	}

	settings, err := db.GetDecryptedSettings(pool)
	if err != nil {
		log.Printf("[Debug] Erro ao obter settings para assinatura MP: %v", err)
		return err
	}

	var secret string
	if settings != nil {
		if secVal, exists := settings.MercadoPagoConfig["webhook_secret"]; exists {
			secret, _ = secVal.(string)
		}
	}

	isIPN := queryParams["topic"] != "" || body.Topic != ""
	if secret != "" && xSignature != "" && xRequestId != "" && !isIPN {
		if dataID == "" {
			return errors.New("id de pagamento ausente na assinatura")
		}

		if !ValidateMPSignature(secret, xSignature, xRequestId, dataID) {
			return errors.New("assinatura do webhook invalida")
		}
	}

	if body.Type != "payment" && body.Topic != "payment" {
		if body.Data == nil || body.Data.ID == "" {
			log.Println("[Debug] Ignorando evento que nao e de pagamento")
			return nil
		}
	}

	paymentID := dataID
	if paymentID == "" {
		log.Println("[Debug] Sem ID de pagamento no payload do webhook")
		return nil
	}

	log.Printf("[Debug] Buscando venda no banco para paymentID: %s", paymentID)
	sale, err := db.FindByPaymentID(pool, paymentID)
	if err != nil {
		return err
	}

	if sale == nil {
		log.Printf("[Debug] Venda nao encontrada por paymentID. Buscando por external reference (ID): %s", paymentID)
		sale, err = db.FindByID(pool, paymentID)
		if err != nil {
			return err
		}
	}

	if sale == nil {
		log.Printf("[Debug] Erro: Venda nao encontrada no sistema para o paymentID: %s", paymentID)
		return fmt.Errorf("venda nao encontrada no sistema: %s", paymentID)
	}

	settings, err = db.GetDecryptedSettings(pool)
	if err != nil {
		log.Printf("[Debug] Erro ao obter settings para API MP: %v", err)
		return err
	}

	var accessToken string
	if settings != nil {
		if tokVal, exists := settings.MercadoPagoConfig["production_access_token"]; exists {
			accessToken, _ = tokVal.(string)
		}
	}

	if accessToken == "" {
		log.Println("[Debug] Erro: access token do MercadoPago nao configurado")
		return errors.New("access token nao configurado")
	}

	log.Printf("[Debug] Chamando API do MercadoPago para buscar status real do pagamento: %s", paymentID)
	mpStatus, err := GetPaymentStatus(accessToken, paymentID)
	if err != nil {
		return err
	}

	newStatus := "pending"
	if mpStatus == "approved" {
		newStatus = "approved"
	} else if mpStatus == "rejected" || mpStatus == "cancelled" {
		newStatus = "failed"
	}

	if sale.Status != newStatus {
		if sale.PaymentDetails == nil {
			sale.PaymentDetails = make(map[string]interface{})
		}
		sale.PaymentDetails["paymentId"] = paymentID
		sale.PaymentDetails["status"] = mpStatus

		log.Printf("[Debug] Atualizando venda no banco para o novo status: %s", newStatus)
		err = db.UpdateSaleStatus(pool, sale.ID, newStatus, sale.PaymentDetails)
		if err != nil {
			return err
		}

		if newStatus == "approved" {
			log.Println("[Debug] Pagamento aprovado! Acionando entrega automatica de produtos.")
			_ = TriggerInternalDelivery(sale.ID)
		} else if newStatus == "failed" {
			log.Println("[Debug] Pagamento falhou! Liberando estoque do produto.")
			_ = db.ReleaseStock(pool, sale.ProductID, sale.Quantity)
		}
	}

	return nil
}

func ProcessEfi(pool *pgxpool.Pool, bodyBytes []byte, headers map[string]string) error {
	log.Println("[Debug] ProcessEfi acionado no Go")

	if pool == nil {
		log.Println("[Debug] Erro: Pool de banco de dados nulo no ProcessEfi")
		return errors.New("pool de banco de dados nulo")
	}

	clientVerify := headers["x-ssl-client-verify"]
	if clientVerify == "" {
		clientVerify = headers["ssl-client-verify"]
	}

	if clientVerify != "" && clientVerify != "SUCCESS" {
		log.Printf("[Debug] Falha na validacao mTLS da Efi no proxy reverso: %s", clientVerify)
		return errors.New("falha na validacao mtls da Efi")
	}

	var body EfiWebhookBody
	if err := json.Unmarshal(bodyBytes, &body); err != nil {
		log.Printf("[Debug] Erro ao decodificar webhook Efi: %v", err)
		return err
	}

	if len(body.Pix) == 0 {
		log.Println("[Debug] Sem itens de Pix no payload do webhook")
		return nil
	}

	settings, err := db.GetDecryptedSettings(pool)
	if err != nil {
		log.Printf("[Debug] Erro ao carregar settings para validar Pix: %v", err)
		return err
	}

	if settings == nil {
		return errors.New("configuracoes da Efi nao localizadas no banco")
	}

	var clientID, clientSecret, certBase64 string
	isProd := false
	if settings.EfiConfig != nil {
		if modeVal, exists := settings.EfiConfig["mode"]; exists {
			isProd = modeVal.(string) == "production"
		}
		if isProd {
			if cidVal, exists := settings.EfiConfig["production_client_id"]; exists {
				clientID, _ = cidVal.(string)
			}
			if csecVal, exists := settings.EfiConfig["production_client_secret"]; exists {
				clientSecret, _ = csecVal.(string)
			}
		} else {
			if cidVal, exists := settings.EfiConfig["sandbox_client_id"]; exists {
				clientID, _ = cidVal.(string)
			}
			if csecVal, exists := settings.EfiConfig["sandbox_client_secret"]; exists {
				clientSecret, _ = csecVal.(string)
			}
		}

		if certVal, exists := settings.EfiConfig["certificate"]; exists {
			certBase64, _ = certVal.(string)
		} else if certVal, exists := settings.EfiConfig["production_certificate_base64"]; exists {
			certBase64, _ = certVal.(string)
		}
	}

	for _, pixItem := range body.Pix {
		txid := pixItem.Txid
		if txid == "" {
			continue
		}

		log.Printf("[Debug] Processando Pix txid: %s", txid)
		sale, err := db.FindByTxid(pool, txid)
		if err != nil {
			log.Printf("[Debug] Erro ao buscar venda por txid %s: %v", txid, err)
			continue
		}

		if sale == nil {
			log.Printf("[Debug] Venda nao encontrada para txid: %s", txid)
			continue
		}

		if sale.Status == "approved" {
			log.Printf("[Debug] Venda txid %s ja esta aprovada. Ignorando.", txid)
			continue
		}

		log.Printf("[Debug] Consultando status real do Pix txid %s na API da Efí", txid)
		efiStatus, err := GetPixStatus(clientID, clientSecret, certBase64, isProd, txid)
		if err != nil {
			log.Printf("[Debug] Erro ao consultar Pix na Efi para txid %s: %v", txid, err)
			continue
		}

		if efiStatus == "CONCLUIDA" && sale.Status != "approved" {
			if sale.PaymentDetails == nil {
				sale.PaymentDetails = make(map[string]interface{})
			}
			sale.PaymentDetails["status"] = "CONCLUIDA"

			log.Printf("[Debug] Pagamento Pix aprovado! Atualizando status no banco para a venda: %s", sale.ID)
			err = db.UpdateSaleStatus(pool, sale.ID, "approved", sale.PaymentDetails)
			if err != nil {
				log.Printf("[Debug] Erro ao atualizar status da venda %s: %v", sale.ID, err)
				continue
			}

			log.Println("[Debug] Pix aprovado com sucesso! Acionando entrega automatica de produtos.")
			_ = TriggerInternalDelivery(sale.ID)
		}
	}

	return nil
}
